### ISSUE TITLE
Unable to Quick Debug

### ISSUE THREAD

**Issue by joshdorsey** [microsoft/vscode-cmake-tools#1698]
### Brief Issue Summary

On two separate computers I'm using practically the same combination of tools but one is able to [quick debug](https://vector-of-bool.github.io/docs/vscode-cmake-tools/debugging.html#quick-debugging) and another (the one I care about) is not able to:
- CMake (3.19.6 on the one that doesn't work, 3.17.0-rc2 on the one that does)
- Clang 10.0.0
- VSCode 1.53.2
- Visual Studio 2019 /w Build Tools installed

The documentation mentions quick debugging for all of 3-4 sentences and it seems to be a CMake Tools thing.  I've installed the `cpptools` extension, and I've checked everything I can think of between these two machines.  My question here is what exactly needs to be in place for this to work, so I can begin to figure out how to get it working.

### Expected:

1. Click the debug button on the bottom bar without a launch.json file
2. Watch CMake build target, ending with a line like
`[build] Build finished with exit code 0`
2. Debugger comes up, and debug console shows a notice referencing the C/C++ Extension for VSCode

### Apparent Behavior:

1. Click the debug button on the bottom bar without a launch.json file
2. Watch CMake build target, but it has lines underneath it looking for a debugger.
```
[proc] Executing command: "C:/Program Files/LLVM/bin/lldb-mi.exe" --version
[proc] Executing command: c:\Users\<me>\.vscode\extensions\ms-vscode.cpptools-1.2.2\debugAdapters\lldb-mi\bin\lldb-mi --version
[proc] Executing command: "C:/Program Files/LLVM/bin/gdb.exe" --version
[proc] Executing command: "C:/Program Files/LLVM/bin/lldb.exe" --version
[proc] Executing command: "C:\Program Files\LLVM\bin\gdb.exe" --version
[proc] Executing command: gdb --version
```
3. Error dialog comes up saying:
    `Unable to find GDB in default search path and gdb.`

This is confusing since the working machine didn't even look for gdb, and found the debugger from `cpptools` just fine.

### Platform and Versions

- **Operating System**: Windows
- **CMake Version**: 3.19.6
- **VSCode Version**: 1.53.2
- **CMake Tools Extension Version**: 1.6.0
- **Compiler/Toolchain**: Clang 10.0.0

### Other Information
The computer that doesn't work has a `settings.json` file with unrelated settings in it (SSH/Python/LaTeX), and the computer that does has no `settings.json` file at all for my user.  On both computers the project doesn't have a `.vscode` folder, so there's not much I can comb through there.  I would provide more specifics about my project, but it's the same between both computers where one works and one doesn't so the project isn't the issue.

The documentation on Quick Debugging seems to be wrong in this case, or at least incomplete, so that's why I'm making an issue for it.

**Comment by andreeis** [comment]
The only setting that I can think of being able to influence this is "cmake.debugConfig". Is that present in your user settings.json on the computer where debugging fails?

Also, I am surprised that you don't have a .vscode subfolder in your projects. Didn't you set the IntelliSense configuration provider to CMake Tools for the CppTools extension to use? Do you set up IntelliSense in a different way, c_cpp_properties.json or compile_commands.json?

**Comment by joshdorsey** [comment]
Thanks for the response,
The only cmake property in my appdata settings.json on the computer that fails is `cmake.configureOnOpen`.

As for IntelliSense, it's a mess.  I'm still in the process of setting up the project, and there's nothing in my .vscode folder in the repository, so I was deleting it locally to reduce the search space for this issue.

I can look at a log file or something else if it would help, what info can I provide?

**Comment by bobbrow** [comment]
> This is confusing since the working machine didn't even look for gdb, and found the debugger from cpptools just fine.

Are both of your machines Windows-based machines?  cpptools does not distribute lldb-mi on Windows, so I'd be really surprised if CMake Tools found that on the working machine if it's a Windows machine.

If you are able to configure your project with CMake Tools, you should be able to get IntelliSense set up relatively quickly by using the `C/C++: Change Configuration Provider` command in VS Code (assuming it wasn't automatically set up for you).  If you have any trouble with that, please let us know as we do expect the IntelliSense configuration integration with cpptools to work once you've successfully configured your project with CMake.

**Comment by joshdorsey** [comment]
Both machines are windows machines (they both have visual studio installed as well).  I should specify, I don't actually have a preference for which debugger to use unless there's some kind of difference, on the machine that works I'm able to pause and examine everything, so it doesn't matter to me.
On the machine that works there's a `vsdbg` and `bin` folder in `%userprofile%\.vscode\extensions\ms-vscode.cpptools-1.2.2\debugAdapters\`, so I assume it's using that.  lldb-mi didn't come with my LLVM binary (although there is something called lldb-vscode.exe).  lldb-mi seems really hard to find on windows from googling around.  I haven't done anything special with any of the lldb variants on either machine though.
The broken machine also has the `vsdbg` folder, and that's why I was looking into maybe my installed visual studio modules having some kind of impact.

Thanks for the IntelliSense advice, as well.

**Comment by joshdorsey** [comment]
Bumped the log level to TRACE on both machines.  The one that worked had a line like the one you predicted, `{..., "config": {"type": "cppvsdbg", ...}}`.  The one that doesn't work has the following output:
```
[build] Build finished with exit code 0
[cache] Reading CMake cache file <project directory>/build/CMakeCache.txt
[cache] Parsing CMake cache string
[proc] Executing command: "C:/Program Files/LLVM/bin/lldb-mi.exe" --version
```
After reading that cache file it went straight to trying to hunt down lldb.

The one that worked made a note: `[main] Debug configuration from cache:` and listed the cppvsdbg configuration.  I looked in the cache file and couldn't find anything overly helpful, but I've deleted my `build` folders on both machines and started from scratch so I'm not sure how useful this is.  Searching around that CMake cache file on both machines for terms like `vsdbg` and `debug` didn't reveal anything obvious.

**Comment by bobbrow** [comment]
cppvsdbg means that the Visual Studio debugger is being used.  Is your code being built by Clang or the Visual Studio compiler on that machine?

**Comment by joshdorsey** [comment]
The active kit is clang, but the log doesn't contain actual build command.  Isn't this behavior expected though?  The docs say that quick debugging (debugging without a launch.json, like I'm trying to do) only works with the `vscode-cpptools` debugger.  I grabbed the CMakeCache.txt file from both machines and ran them through vimdiff and it looks like the broken one is grabbing a bunch of executables from mingw-w64 even though I have clang selected as the active kit, so I'm looking into that right now.

**Comment by bobbrow** [comment]
> The docs say that quick debugging (debugging without a launch.json, like I'm trying to do) only works with the `vscode-cpptools` debugger.

Sorry for the confusion.  When I said the "Visual Studio Debugger" I was referring to the fact that vscode-cpptools redistributes and uses the Visual Studio debugger implementation when `cppvsdbg` is set as the debug type.  lldb-mi is selectable only when `cppdbg` is set as the debug type.  So something in your configuration is causing the extension to go down that path instead of the `cppvsdbg` path.

When you configure the project what is printed out as the C and CXX compiler for each machine?  You may need to run the `CMake: Delete Cache and Reconfigure` command to get it to print out.  However, as I was testing this, I was seeing something weird when I selected a Clang kit and it turned out I had the `cmake.generator` setting set to "Visual Studio 15 2017" instead of "Ninja" which was causing some problems.  So that's something else you can check.  (@andreeis we should open a bug to notify the user or just fix it on the fly.)

Your log should look something like this:
```
[cmakefileapi-driver] Invoking CMake C:\Program Files\CMake\bin\cmake.EXE with arguments ["--no-warn-unused-cli","-DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE","-DCMAKE_BUILD_TYPE:STRING=Debug","-DCMAKE_C_COMPILER:FILEPATH=C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Preview\\VC\\Tools\\Llvm\\bin\\clang-cl.exe","-DCMAKE_CXX_COMPILER:FILEPATH=C:\\Program Files (x86)\\Microsoft Visual Studio\\2019\\Preview\\VC\\Tools\\Llvm\\bin\\clang-cl.exe","-Hd:/test/cmaketest","-Bd:/test/cmaketest/build","-G","Ninja"]
[proc] Executing command: "C:\Program Files\CMake\bin\cmake.EXE" --no-warn-unused-cli -DCMAKE_EXPORT_COMPILE_COMMANDS:BOOL=TRUE -DCMAKE_BUILD_TYPE:STRING=Debug "-DCMAKE_C_COMPILER:FILEPATH=C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\Llvm\bin\clang-cl.exe" "-DCMAKE_CXX_COMPILER:FILEPATH=C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\Llvm\bin\clang-cl.exe" -Hd:/test/cmaketest -Bd:/test/cmaketest/build -G Ninja
[cmake] Not searching for unused variables given on the command line.
[kit] OK running C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Auxiliary\Build\vcvarsall.bat x86, env vars: [...]
[cmake] -- The C compiler identification is Clang 11.0.0 with MSVC-like command-line
[cmake] -- The CXX compiler identification is Clang 11.0.0 with MSVC-like command-line
[cmake] -- Detecting C compiler ABI info
[cmake] -- Detecting C compiler ABI info - done
[cmake] -- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Preview/VC/Tools/Llvm/bin/clang-cl.exe - skipped
[cmake] -- Detecting C compile features
[cmake] -- Detecting C compile features - done
[cmake] -- Detecting CXX compiler ABI info
[cmake] -- Detecting CXX compiler ABI info - done
[cmake] -- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Preview/VC/Tools/Llvm/bin/clang-cl.exe - skipped
[cmake] -- Detecting CXX compile features
[cmake] -- Detecting CXX compile features - done
[cmake] -- Configuring done
[cmake] -- Generating done
```

**Comment by andreeis** [comment]
Opened https://github.com/microsoft/vscode-cmake-tools/issues/1702 to track such incompatibilities between the kit and the user defined generator.

**Comment by joshdorsey** [comment]
Both machines say:
```
[cmake] -- The C compiler identification is Clang 10.0.0 with GNU-like command-line
[cmake] -- The CXX compiler identification is Clang 10.0.0 with GNU-like command-line
```
(one says Clang 11.0.0, but I updated yesterday and it hasn't affected anything)

The difference in logging produced with the `CMake: Delete Cache and Reconfigure` command on both machines is that the working one has these lines (for C and CXX):
```
[cmake] -- Check for working C compiler: C:/Program Files/LLVM/bin/clang.exe
[cmake] -- Check for working C compiler: C:/Program Files/LLVM/bin/clang.exe - works
```
and the broken one has:
`[cmake] -- Check for working C compiler: C:/Program Files/LLVM/bin/clang.exe - skipped`

**Comment by andreeis** [comment]
Is the clang.exe that is skipped (on the broken environment) in good state? Does it work when you run it from terminal?
I wonder if a clang reinstallation over there may help.

**Comment by joshdorsey** [comment]
The build has worked absolutely fine this whole time on both machines, debugging is the only issue.  I did verify that my Clang 11.0.0 install compiles a simple test program fine as well though.  I don't know enough about CMake to know what it means for Clang to skip the check like that.

**Comment by andreeis** [comment]
Ok, no idea for now, brainstorming... It might help if you share a repro project with us.

**Comment by joshdorsey** [comment]
The project itself is invariant between the working/broken machines so I wasn't thinking about that.  I was able to recreate this behavior with a really simple project:
./src/main.cpp:
```
#include <iostream>
int main() {
    std::cout << "This is a test\n";
}
```
./src/CMakeLists.txt
```
set(CMAKE_CXX_STANDARD 17)
set(CMAKE_CXX_STANDARD_REQUIRED True)

# The project I'm actually trying to debug uses SDL2 so I figured I'd leave this in
find_package(SDL2 REQUIRED)

add_executable(test main.cpp)
target_include_directories(test PUBLIC ${SDL2_INCLUDE_DIR})
target_link_libraries(test SDL2::SDL2-static)
```
./CMakeLists.txt:
```
cmake_minimum_required(VERSION 3.10)
project(Test VERSION 0.1)
add_subdirectory(src)
```

From what I gather, the best theory so far is that when CMake finds some mingw executables on my system and for whatever reason even though it knows my toolchain is Clang those end up in the CMakeCache.txt file in my build root[1], VSCode gets confused about how to debug the project.

[1] I've tried putting my LLVM binaries at a higher precedence than the mingw folder in my %PATH% (see [this](https://github.com/microsoft/vscode-cmake-tools/issues/1698#issuecomment-790134733)), and I've tried deleting the mingw folder from my %PATH% entirely, but CMake could have cached it or something so I can verify that better.

**Comment by andreeis** [comment]
If you delete your build folder and reconfigure, the cache will be recreated so you can verify this way if moving paths around helps or not. Also, to make sure we start fresh with everything, you can run the command "CMake: Reset state for troubleshooting" (after which you will have to select again a kit, a variant, anything you did before to setup your project). Hopefully moving the mingw paths further than LLVM and refreshing your project setup will help to unblock you.
In the meantime, we'll try to use your findings to reproduce on our side and fix the problem.

**Comment by joshdorsey** [comment]
Even with nothing linux-y in my %PATH%, VSCode was still tripping up on what was in the CMakeCache.txt when debugging.  I was able to get it working on both computers by essentially recreating the environment from the working computer on the broken one.  I did this by:
- Uninstalling mingw-w64
- Uninstalling normal MinGW
- Reinstalling MinGW, and the mingw32-make package
- Adding `C:\MinGW\bin` to %PATH%, copying `mingw32-make.exe` and naming it `make.exe`

This process lets me build and debug in VSCode.  If I remove `make.exe`, I'm able to build the project with `mingw32-make.exe` by selecting "MinGW Makefiles" in the cmake GUI, but it no longer works in VSCode (an error message about not knowing what generator to use).

This technically solves my problem, but puts me squarely under the impression that I'm going about using Clang in some strange, non-standard way.  Have I been doing something unusual this whole time?
